File: JavaDataManager/src/joo/server/helpers/MsgHelper.java
package joo.server.helpers;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import joo.databaseManager.BasicModel;
import joo.databaseManager.helpers.SerializerHelper;

public class MsgHelper {

	private final static byte TYPE_OBJ =1;
	private final static byte TYPE_INT=2;
	private final static byte TYPE_STRING=3;
	private final static byte TYPE_DATE=4;
	private final static byte TYPE_BUFFER=5;
	private final static byte TYPE_DOUBLE=6;
	private final static byte TYPE_FLOAT=7;
	
	private int type;
	private List<Object> values; 
	
	public MsgHelper(int type){
		values=new ArrayList<Object>();
		this.type=type;
	}
	public MsgHelper(byte[] buffer){
		values=new ArrayList<Object>();
		ByteBuffer byteBuffer=ByteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN);
		type=byteBuffer.getInt();
		int count=byteBuffer.getInt();
		for (int i = 0; i < count; i++) {
			byte typeValue=byteBuffer.get();
			int sizeValue=byteBuffer.getInt();
			switch (typeValue) {
				case TYPE_OBJ:
				{
					byte[] valueBuffer=new byte[sizeValue];
					byteBuffer.get(valueBuffer);
					
					BasicModel value = SerializerHelper.deserialize(valueBuffer);
					values.add(value);
				}
				break;
				case TYPE_INT:
				values.add(byteBuffer.getInt());
				break;
				case TYPE_DATE:
				{
					byte[] valueBuffer=new byte[sizeValue];
					byteBuffer.get(valueBuffer);
					String strDate=new String(valueBuffer);
					SimpleDateFormat formater = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
					try {
						Date date = formater.parse(strDate);
						values.add(date);
					} catch (ParseException e) {
						e.printStackTrace();
					}
				}
				break;
				case TYPE_STRING:
				{
					byte[] valueBuffer=new byte[sizeValue];
					byteBuffer.get(valueBuffer);
					String value;
					try {
						value = new String(valueBuffer,"UTF-8");
						values.add(value);
					} catch (UnsupportedEncodingException e) {
						e.printStackTrace();
					}
				}
				break;
				case TYPE_BUFFER:
				{
					byte[] valueBuffer=new byte[sizeValue];
					byteBuffer.get(valueBuffer);
					values.add(valueBuffer);
				}
				break;
				case TYPE_DOUBLE:
					values.add(byteBuffer.getDouble());
					break;
				case TYPE_FLOAT:
					values.add(byteBuffer.getFloat());
					break;
				default:
					throw new IllegalArgumentException("Buffer com dados invalidos. Tido de valor não reconhecido");
			}
		}
		System.out.println();
	}
	
	public void addValue(BasicModel obj){
		values.add(obj);
	}
	public void addValue(int value){
		values.add(value);
	}
	public void addValue(String value){
		values.add(value);
	}
	public void addValue(Date value){
		values.add(value);
	}
	public void addValue(byte[] buffer){
		values.add(buffer);
	}
	public void addValue(Double value) {
		values.add(value);
	}
	public void addValue(Float value){
		values.add(value);
	}
	
	public void addValue(Object value){
		values.add(value);
	}
	
	public BasicModel getValue(int index){
		return (BasicModel)values.get(index);
	}
	public int getValueToInt(int index){
		return (int)values.get(index);
	}
	public String getValueToString(int index){
		return (String)values.get(index);
	}
	public Date getValueToDate(int index){
		return (Date)values.get(index);
	}
	public byte[] getValueToBuffer(int index){
		return (byte[])values.get(index);
	}
	public Double getValueToDouble(int index){
		return (Double)values.get(index);
	}
	public Float getValueToFloat(int index){
		return (Float)values.get(index);
	}
	
	public int getType(){
		return type;
	}
	public int getCountValues(){
		return values.size();
	}
	
	public byte[] toArray(){
		try{
			ByteArrayOutputStream out=new ByteArrayOutputStream();
			ByteBuffer typeBuffer=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
			typeBuffer.putInt(type);
			out.write(typeBuffer.array());
			ByteBuffer sizeBuffer=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
			sizeBuffer.putInt(values.size());
			out.write(sizeBuffer.array());
			for (int i = 0; i < values.size(); i++) {
				Object value=values.get(i);
				
				if(BasicModel.class.isInstance(value)){
					out.write(TYPE_OBJ);
					byte[] buffer=SerializerHelper.serializer((BasicModel)value);
					if(buffer==null){
						continue;
					}
					ByteBuffer size=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
					size.putInt(buffer.length);
					out.write(size.array());
					out.write(buffer);
				}else{
					if(String.class.isInstance(value)){
						out.write(TYPE_STRING);
						byte[] buffer=((String)value).getBytes("UTF-8");
						ByteBuffer size=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
						size.putInt(buffer.length);
						out.write(size.array());
						out.write(buffer);
					}else{
						if(Date.class.isInstance(value)){
							out.write(TYPE_DATE);
							SimpleDateFormat formater = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
							byte[] buffer=formater.format((Date)value).getBytes();
							ByteBuffer size=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
							size.putInt(buffer.length);
							out.write(size.array());
							out.write(buffer);
						}else{
							if(int.class.isInstance(value) || Integer.class.isInstance(value)){
								out.write(TYPE_INT);
								 ByteBuffer size=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
								size.putInt(4);
								out.write(size.array());
								ByteBuffer b = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
								b.putInt((int)value);
								out.write(b.array());
							}else{
								if(byte[].class.isInstance(value)){
									byte[] buffer=(byte[])value;
									out.write(TYPE_BUFFER);
									ByteBuffer size=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
									size.putInt(buffer.length);
									out.write(size.array());
									out.write(buffer);
								}else{
									if(Double.class.isInstance(value)) {
										out.write(TYPE_DOUBLE);
										ByteBuffer size=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
										size.putInt(8);
										out.write(size.array());
										ByteBuffer b = ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN);
										b.putDouble((Double)value);
										out.write(b.array());
									} else {									
										if(Float.class.isInstance(value)) {
											out.write(TYPE_FLOAT);
											ByteBuffer size=ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
											size.putInt(4);
											out.write(size.array());
											ByteBuffer b = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
											b.putFloat((Float)value);
											out.write(b.array());
										} else {
											throw new IllegalArgumentException("Valor do tipo "+value.getClass().getName()+" não é permitido.");
										}
									}
								}
							}
						}
					}
				}
			}
			return out.toByteArray();
		}catch(IOException err){
			err.printStackTrace();
			return null;
		}
	}

}

File: NetDataManager/ClientJavaServer/MsgHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using Database;

namespace Server
{
    public class MsgHelper
    {
        public enum TYPES : short
        {
            OBJ =1,
	        INT=2,
	        STRING=3,
	        DATE=4,
	        BUFFER=5,
	        DOUBLE=6,
	        FLOAT=7
        }
	
	    private int type;
	    private List<Object> values; 
	
	    public MsgHelper(int type){
		    values=new List<Object>();
		    this.type=type;
	    }
	    public MsgHelper(byte[] buffer,String assemblyFullName){
		    values=new List<Object>();
            int index=0;
		    type=BitConverter.ToInt32(buffer,index);
            index+=sizeof(int);
		    int count=BitConverter.ToInt32(buffer,index);
            index+=sizeof(int);
		    for (int i = 0; i < count; i++) {
			    byte typeValue=buffer[index];
                index+=1;
			    int sizeValue=BitConverter.ToInt32(buffer,index);
                index+=sizeof(int);
			    switch (typeValue) {
				    case (int)TYPES.OBJ:
				    {
					    byte[] valueBuffer=new byte[sizeValue];
					    Buffer.BlockCopy(buffer,index,valueBuffer,0,sizeValue);
                        index+=sizeValue;
					   
					    BasicModel value = JavaSerializer.Deserialize<BasicModel>(valueBuffer,assemblyFullName);
					    values.Add(value);
				    }
				    break;
                    case (int)TYPES.INT:
				    values.Add(BitConverter.ToInt32(buffer,index));
                    index+=sizeof(int);
				    break;
                    case (int)TYPES.DATE:
				    {
					    byte[] valueBuffer=new byte[sizeValue];
					    Buffer.BlockCopy(buffer,index,valueBuffer,0,sizeValue);
                        index+=sizeValue;
					    String strDate=Encoding.UTF8.GetString(valueBuffer);
					   
						DateTime date = DateTime.Parse(strDate);
						values.Add(date);					   
				    }
				    break;
                    case (int)TYPES.STRING:
				    {
					    byte[] valueBuffer=new byte[sizeValue];
					    Buffer.BlockCopy(buffer,index,valueBuffer,0,sizeValue);
                        index+=sizeValue;
					    String value=Encoding.UTF8.GetString(valueBuffer);
					    values.Add(value);
				    }
				    break;
                    case (int)TYPES.BUFFER:
				    {
					    byte[] valueBuffer=new byte[sizeValue];
					    Buffer.BlockCopy(buffer,index,valueBuffer,0,sizeValue);
                        index+=sizeValue;
					    values.Add(valueBuffer);
				    }
				    break;
                    case (int)TYPES.DOUBLE:
					    values.Add(BitConverter.ToDouble(buffer,index));
					    break;
                    case (int)TYPES.FLOAT:
					    values.Add(BitConverter.ToSingle(buffer,index));
					    break;
				    default:
					    throw new ArgumentException("Buffer com dados invalidos. Tido de valor não reconhecido");
			    }
		    }
	    }
	
	    public void AddValue(BasicModel obj){
		    values.Add(obj);
	    }
	    public void AddValue(int value){
		    values.Add(value);
	    }
	    public void AddValue(String value){
		    values.Add(value);
	    }
	    public void AddValue(DateTime value){
		    values.Add(value);
	    }
	    public void AddValue(byte[] buffer){
		    values.Add(buffer);
	    }
	    public void AddValue(Double value) {
		    values.Add(value);
	    }
	    public void AddValue(float value){
		    values.Add(value);
	    }
        public void AddValue(Decimal value)
        {
            values.Add(value);
        }
	
	    public BasicModel GetValue(int index){
		    return (BasicModel)values[index];
	    }
	    public int GetValueToInt(int index){
		    return (int)values[index];
	    }
	    public String GetValueToString(int index){
		    return (String)values[index];
	    }
	    public DateTime GetValueToDate(int index){
		    return (DateTime)values[index];
	    }
	    public byte[] GetValueToBuffer(int index){
		    return (byte[])values[index];
	    }
	    public Double GetValueToDouble(int index){
		    return (Double)values[index];
	    }
	    public float GetValueToFloat(int index){
		    return (float)values[index];
	    }
	
	    public int GetMsgType(){
		    return type;
	    }
	    public int GetCountValues(){
		    return values.Count;
	    }
	
	    public byte[] ToArray(){
			MemoryStream outStream=new MemoryStream();
			outStream.Write(BitConverter.GetBytes(type),0,sizeof(int));
			outStream.Write(BitConverter.GetBytes(values.Count),0,sizeof(int));
			for (int i = 0; i < values.Count; i++) {
				Object value=values[i];
				
				if(typeof(BasicModel).IsInstanceOfType(value)){
                    outStream.WriteByte((int)TYPES.OBJ);
					byte[] buffer=JavaSerializer.Serializer((BasicModel)value);
					if(buffer==null){
						continue;
					}
					outStream.Write(BitConverter.GetBytes(buffer.Length),0,sizeof(int));
					outStream.Write(buffer,0,buffer.Length);
				}
				if(typeof(String).IsInstanceOfType(value)){
                    outStream.WriteByte((int)TYPES.STRING);
					byte[] buffer=Encoding.UTF8.GetBytes((String)value);
					outStream.Write(BitConverter.GetBytes(buffer.Length),0,sizeof(int));
					outStream.Write(buffer,0,buffer.Length);
				}
				if(typeof(DateTime).IsInstanceOfType(value)){
                    outStream.WriteByte((int)TYPES.DATE);
                    byte[] buffer = Encoding.UTF8.GetBytes(((DateTime)value).ToString("yyyy-MM-dd HH:mm:ss"));
					outStream.Write(BitConverter.GetBytes(buffer.Length),0,sizeof(int));
					outStream.Write(buffer,0,buffer.Length);
				}
				if(typeof(int).IsInstanceOfType(value)){
                    outStream.WriteByte((int)TYPES.INT);
					outStream.Write(BitConverter.GetBytes(sizeof(int)),0,sizeof(int));
					outStream.Write(BitConverter.GetBytes((int)value),0,sizeof(int));
				}
				if(typeof(byte[]).IsInstanceOfType(value)){
					byte[] buffer=(byte[])value;
                    outStream.WriteByte((int)TYPES.BUFFER);
					outStream.Write(BitConverter.GetBytes(buffer.Length),0,sizeof(int));
					outStream.Write(buffer,0,buffer.Length);
				}
				if(typeof(Double).IsInstanceOfType(value)) {
                    outStream.WriteByte((int)TYPES.DOUBLE);
					outStream.Write(BitConverter.GetBytes(sizeof(Double)),0,sizeof(int));
					outStream.Write(BitConverter.GetBytes((Double)value),0,sizeof(Double));
				} 								
				if(typeof(float).IsInstanceOfType(value)) {
                    outStream.WriteByte((int)TYPES.FLOAT);
					outStream.Write(BitConverter.GetBytes(sizeof(float)),0,sizeof(int));
					outStream.Write(BitConverter.GetBytes((float)value),0,sizeof(float));
				}
                if (typeof(Decimal).IsInstanceOfType(value))
                {
                    outStream.WriteByte((int)TYPES.FLOAT);
                    outStream.Write(BitConverter.GetBytes(sizeof(float)), 0, sizeof(int));
                    outStream.Write(BitConverter.GetBytes((float)value), 0, sizeof(float));
                }
			}
			return outStream.ToArray();
	    }
    }
}
